File: Leetcode/Array&String/31_next-permutation/main.cpp
// 31. 下一个排列
// Created by bob on 2022/6/16.
//

#include<vector>
#include "KPrint.h"

using namespace std;

class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int i = nums.size() - 2;
        while (i>=0 && nums[i-1] >= nums[i]) {
            i--;
        }
        if (i >=0) {
            int j = nums.size() - 1;
            while (j>=0 && nums[i] >= nums[j]) {
                j--;
            }
            swap(nums[i], nums[j]);
        }
        reverse(nums.begin() + i + 1, nums.end());
    }
};

int main() {
    vector<int> nums = {1, 2, 3};
    Solution().nextPermutation(nums);
    printVector(nums);
}

File: Leetcode/Array&String/31_next-permutation/main.py
from typing import List


class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        i = len(nums) - 2
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1

        if i >= 0:
            j = len(nums) - 1
            while j >= 0 and nums[i] >= nums[j]:
                j -= 1
            nums[i], nums[j] = nums[j], nums[i]

        l, r = i + 1, len(nums) - 1
        while l < r:
            nums[l], nums[r] = nums[r], nums[l],
            l += 1
            r -= 1


if __name__ == '__main__':
    nums = [4, 5, 2, 6, 3, 1]
    Solution().nextPermutation(nums)
    print(nums)  # expect: [4,5,3,1,2,6]
